File: backend/src/main/java/com/kratosgado/chatapp/auth/AuthService.java

package com.kratosgado.chatapp.auth;

import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import com.kratosgado.chatapp.auth.dto.RegisterDto;
import com.kratosgado.chatapp.services.JwtService;
import com.kratosgado.chatapp.users.User;
import com.kratosgado.chatapp.users.UserRepo;
import com.kratosgado.chatapp.utils.ApiException;
import com.kratosgado.chatapp.utils.UtilConstants;

import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class AuthService {

  private final UserRepo userRepo;
  private final PasswordEncoder passwordEncoder;
  private final JwtService jwtService;

  public User register(RegisterDto dto, HttpServletResponse res) {
    if (userRepo.existsByEmail(dto.email())) {
      ApiException.badRequest("Email already registered");
    }

    if (!dto.password().equals(dto.confirmPassword())) {
      ApiException.badRequest("Passwords do not match");
    }

    User user = new User();
    user.setName(dto.name());
    user.setEmail(dto.email());
    user.setPassword(passwordEncoder.encode(dto.password()));
    userRepo.save(user);

    String token = jwtService.generateToken(user);
    Cookie cookie = new Cookie("jwt", token);

    cookie.setHttpOnly(true);
    cookie.setSecure(true);
    cookie.setPath("/");
    cookie.setMaxAge(UtilConstants.ONE_DAY_IN_SECONDS);
    res.addCookie(cookie);
    return user;
  }

}

File: backend/src/main/kotlin/com/kratosgado/chatapp/auth/AuthService.kt
package com.kratosgado.chatapp.auth

import com.kratosgado.chatapp.auth.dto.LoginDto
import com.kratosgado.chatapp.auth.dto.LoginResponseDto
import com.kratosgado.chatapp.auth.dto.RegisterDto
import com.kratosgado.chatapp.auth.dto.UserDto
import com.kratosgado.chatapp.services.JwtService
import com.kratosgado.chatapp.users.User
import com.kratosgado.chatapp.users.UserRepo
import com.kratosgado.chatapp.utils.ApiException
import com.kratosgado.chatapp.utils.UtilConstants
import jakarta.servlet.http.Cookie
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class AuthService(
    private val userRepo: UserRepo,
    private val passwordEncoder: PasswordEncoder,
    private val jwtService: JwtService,
) {
    fun register(dto: RegisterDto): User {
        if (userRepo.existsByEmail(dto.email)) {
            throw ApiException.badRequest("Email already registered")
        }

        if (dto.password != dto.confirmPassword) {
            throw ApiException.badRequest("Passwords do not match")
        }

        val user = User()
        user.name = dto.name
        user.email = dto.email
        user.password = passwordEncoder.encode(dto.password)
        userRepo.save(user)

        val token = jwtService.generateToken(user)
        return user
    }

    fun login(dto: LoginDto): LoginResponseDto {
        val user = userRepo.findByEmail(dto.email) ?: throw ApiException.badRequest("Invalid email or password")

        if (!passwordEncoder.matches(dto.password, user.password)) {
            throw ApiException.badRequest("Invalid email or password")
        }

        val token = jwtService.generateToken(user)
        val cookie = Cookie("jwt", token)

        return LoginResponseDto(token, UserDto.fromEntity(user))
    }

    fun logout(res: HttpServletResponse) {
        val cookie = Cookie("jwt", "")
        cookie.isHttpOnly = true
        cookie.secure = true
        cookie.path = "/"
        cookie.maxAge = 0
        res.addCookie(cookie)
    }
}
